src/main/scala/akka/projection/testing/dynamodb/DynamoDBProjectionHandler.scala

```diff
@@ -0,0 +1,151 @@
+      log.info(
+        "Projection [{}] throughput [{}] events/s in [{}] ms",
+        projectionId.id,
+        1000 * count / durationMs,
+        durationMs)
+      count = 0
+      startTime = System.nanoTime()
+    }
+
+    if (readOnly)
+      Future.successful(Nil)
+    else {
+      val attributes = Map(
+        Events.Attributes.Id -> AttributeValue.fromS(s"${envelope.event.testName}-$projectionIndex"),
+        Events.Attributes.Event -> AttributeValue.fromS(envelope.event.payload)).asJava
+
+      Future.successful(
+        Seq(
+          TransactWriteItem.builder
+            .put(
+              Put.builder
+                .tableName(Events.TableName)
+                .item(attributes)
+                .build())
+            .build()))
+    }
+  }
+}
+
+// when using this consider reducing failure otherwise a high chance of at least one grouped envelope
+// causing an error and no progress will be made
+class DynamoDBGroupedProjectionHandler(
+    projectionId: ProjectionId,
+    projectionIndex: Int,
+    readOnly: Boolean,
+    failEvery: Int,
+    client: DynamoDbAsyncClient)
+    extends Handler[Seq[EventEnvelope[ConfigurablePersistentActor.Event]]] {
+
+  import DynamoDBTables.Events
+
+  private val log: Logger = LoggerFactory.getLogger(getClass)
+  private var startTime = System.nanoTime()
+  private var count = 0
+
+  override def process(envelopes: Seq[EventEnvelope[ConfigurablePersistentActor.Event]]): Future[Done] = {
+    log.trace("Projection [{}] processing group of [{}] events", projectionId.id, envelopes.size)
+
+    envelopes.foreach { envelope =>
+      if (failEvery != Int.MaxValue && Random.nextInt(failEvery) == 1) {
+        throw new RuntimeException(
+          s"Simulated failure when processing persistence id [${envelope.persistenceId}]," +
+            s" sequence nr [${envelope.sequenceNr}], offset [${envelope.offset}]") with NoStackTrace
+      }
+    }
+
+    count += envelopes.size
+    if (count >= 1000) {
+      val durationMs = (System.nanoTime() - startTime) / 1000 / 1000
+      log.info(
+        "Projection [{}] throughput [{}] events/s in [{}] ms",
+        projectionId.id,
+        1000 * count / durationMs,
+        durationMs)
+      count = 0
+      startTime = System.nanoTime()
+    }
+
+    if (readOnly)
+      Future.successful(Done)
+    else {
+      // TODO: multiple batches if groups are larger than 25 (hard limit on batch size)
+
+      val items = envelopes.map { envelope =>
+        val attributes = Map(
+          Events.Attributes.Id -> AttributeValue.fromS(s"${envelope.event.testName}-$projectionIndex"),
+          Events.Attributes.Event -> AttributeValue.fromS(envelope.event.payload)).asJava
+
+        WriteRequest.builder
+          .putRequest(
+            PutRequest.builder
+              .item(attributes)
+              .build())
+          .build()
+      }.asJava
+
+      client
+        .batchWriteItem(
+          BatchWriteItemRequest.builder
```

Comment:
This is not checking the unprocessedItems so it's suffering from the same problem as https://github.com/akka/akka-projection/pull/1260

I also created issue https://github.com/akka/akka-projection/issues/1286

Reply:
Good point. Using a provided retry utility sounds good. Will add this.

Reply:
https://github.com/akka/akka-projection/pull/1287

Reply:
Updated to the retry utility.